File: src/main/java/de/schramm/royalbash/controller/service/core/effect/DiscardHandcardsEffect.java
package de.schramm.royalbash.controller.service.core.effect;

import de.schramm.royalbash.controller.service.core.Context;
import de.schramm.royalbash.controller.service.core.Game;
import lombok.Builder;
import lombok.Value;
import lombok.val;

@Value
@Builder
public class DiscardHandcardsEffect {

    private final int amountOfCards;

    public Game invoke(Context context) {

        val game = context.getGame();
        val target = context.getTargetPlayer();

        return context.getTargetPlayer()
                .flatMap(targetPlayer -> game.findPlayer(targetPlayer)
                        .map(player -> player.discardCards(amountOfCards))
                        .map(player -> game.updatePlayer(targetPlayer, player)))
                .orElse(game);
    }
}

File: src/main/kotlin/de/schramm/royalbash/controller/service/core/effect/DiscardHandcardsEffect.kt
package de.schramm.royalbash.controller.service.core.effect

import de.schramm.royalbash.controller.service.core.Context
import de.schramm.royalbash.controller.service.core.Game
import de.schramm.royalbash.controller.service.core.Player

data class DiscardHandcardsEffect(val amountOfCards: Int) {

    operator fun invoke(context: Context): Game {

        val game = context.game

        return context.getTargetPlayer()
                .flatMap { targetPlayer ->
                    game.findPlayer(targetPlayer)
                            .map<Player> { player -> player.discardCards(amountOfCards) }
                            .map { player -> game.updatePlayer(targetPlayer, player) }
                }
                .orElse(game)
    }
}
